tor-browser: 12.5.6 -> 13.0

felschr opened the pull request:
## Description of changes
https://blog.torproject.org/new-release-tor-browser-130/

Equivalent Mullvad Browser PR: #256935.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

felschr committed:
tor-browser: adjust update script for upcoming changes

Both Tor Browser & Mullvad Browser will use slightly different archive
names beginnign with release 13.0. In addition to adjusting to those
changes, the update script now supports prereleases via an argument.

felschr committed:
tor-browser: 12.5.6 -> 13.0

https://blog.torproject.org/new-release-tor-browser-130/

Also removes some workarounds for Tor sockets which are no longer
needed.

felschr committed:
tor-browser: update licenses

To match licenses with mullvad-browser.

panicgh commented on pkgs/applications/networking/browsers/mullvad-browser/default.nix:
```diff
@@ -228,7 +228,7 @@ stdenv.mkDerivation rec {
       inherit pname version meta;
       baseUrl = "https://cdn.mullvad.net/browser/";
-      prefix = "mullvad-browser-";
+      name = "mullvad-browser";
```
This doesn't belong to this tor-browser PR.

felschr replied:
Well, I changed the argument in the update script, so this change is needed for it not to break.

panicgh replied:
Ah, now I see. Sorry for the noise.

felschr committed:
tor-browser: reorder source urls

dist.torproject.org only ever contains the latest release, while
releases on archive.torproject.org should be available forever.

panicgh commented on pkgs/applications/networking/browsers/tor-browser/default.nix:
```diff
@@ -99,8 +99,8 @@ lib.warnIf (useHardenedMalloc != null)
       urls = [
-        "https://dist.torproject.org/torbrowser/${version}/tor-browser-linux64-${version}_${lang}.tar.xz"
         "https://archive.torproject.org/tor-package-archive/torbrowser/${version}/tor-browser-linux64-${version}_${lang}.tar.xz"
+        "https://dist.torproject.org/torbrowser/${version}/tor-browser-linux64-${version}_${lang}.tar.xz"
```
I'm not sure if it is nice of us to hit archive.torproject.org first. It might be a slower machine that gets easier overloaded than the official "primary" machine for distribution. On the flip side, dist.torproject.org removes old releases, so if an old Tor Browser is to be built, it initially fails and tries again with the archive. I would prefer that over requesting everything directly from the archive.

felschr replied:
Well, dist only ever has the latest release. Also 13.0 was released on archive before dist.
Additionally I talked with a Tor Browser dev recently (regarding the Mullvad Browser releases) and got the following statement:
> More generally, you should not use [dist.torproject.org](http://dist.torproject.org/) in automation since we only typically keep
the latest build on that server, whereas [archive.torproject.org](http://archive.torproject.org/) URLs are stable and should be
available forever into the future.

panicgh commented on pkgs/applications/networking/browsers/tor-browser/update.nix:
```diff
@@ -47,13 +50,13 @@ writeShellScript "update-${pname}" ''
   for platform in ${lib.escapeShellArgs meta.platforms}; do
     arch="''${platforms[$platform]}"
-    sha256=$(cat "$HOME/shasums" | grep "${prefix}""$arch-$version""${suffix}" | cut -d" " -f1)
+    sha256=$(cat "$HOME/shasums" | grep "${name}-$arch-$version.tar.xz" | cut -d" " -f1)
```
`cat` seems unnecessary.

```suggestion
    sha256=$(grep "${name}-$arch-$version.tar.xz" "$HOME/shasums" | cut -d" " -f1)
```

panicgh reviewed (commented):
We may remove a workaround that is still in place for cleaning up a directory, see https://github.com/NixOS/nixpkgs/blob/808279954b08bb08e419e0ef3b78f2f51d36d3ca/pkgs/applications/networking/browsers/tor-browser/default.nix#L338-L340
This seems to be not necessary any more. I removed it locally and restarted Tor Browser several times and it never failed to connect with the socket (the directory in the `network.proxy.socks` about:config option is simply incremented each time). The upstream blog post on the release also mentions that the "red screen of death" has been removed and internals reworked.

The workaround was introduced due to seeing the red screen of death after Tor Browser is restarted, see https://github.com/NixOS/nixpkgs/pull/205463#issuecomment-1436064480.

panicgh commented:
You may also mention the blog post URL in the message of the version bump commit.

felschr commented:
> the directory in the `network.proxy.socks` about:config option is simply incremented each time

Could you clarify that? We're actually overwriting `network.proxy.socks` currently because of https://github.com/NixOS/nixpkgs/issues/219750 which was implemented in https://github.com/NixOS/nixpkgs/pull/226177.

I've just removed the `user_pref` setting that config from our package and it seems to work again.
Probably this was also covered by the "red screen of death" fixes in 13.0.

felschr commented:
It seemed that `network.proxy.socks` remained `file:///run/user/1000/Tor/socks.socket` after removing that option, even after resetting that value in `about:config` and restarting the browser.

I've reset some other related values in `about:config` and now it gets set to `file:///run/user/1000/Tor-5/socks.socket`, though.

felschr commented:
I've just replicated an update from 12.5.6 with existing state to 13.0 again and now it gets set correctly to `file:///run/user/1000/Tor-1/socks.socket` after the update.
Seems to have been some issue on my side, or perhaps I was just blind 😅.

panicgh reviewed (approved):
I've used it throughout the day, started Tor Browser multiple times, and had no issues.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1156
